### DISCUSSION TITLE
Question regarding using a custom closure/anonymous function for where

### DISCUSSION THREAD

**Discussion by piperun** [simolus3/drift#3167] [Q&A] [answered]
Hi! I'm trying to refactor my code and I'm in the process of refactoring my batch functions, insert & delete works without any issues but I've stumble upon an issue when it comes to using `batch.update`:
```error
The getter 'id' isn't defined for the type 'Table'.
Try importing the library that defines 'id', 
correcting the name to the name of an existing getter, or defining a getter or field named 'id'
```
Usage:
```dart
batch.updateBatch(folders, folderUpdate, (t) => t.id.equals(folderId));
```

The issue being that since I'm taking in a Table, I assume that `.where` doesn't have enough context to be able to know which table it is (despite said table being included) is there something I am missing?

Implementation:

```dart
  void updateBatch(TableInfo table, List<Insertable> rows,
      Expression<bool> Function(Table) where) async {
    await _db.batch((batch) async {
      for (final row in rows) {
        batch.update(table, row, where: where);
      }
    });
  }
```

**Comment by simolus3** [comment, marked as answer]
> The issue being that since I'm taking in a Table, I assume that `.where` doesn't have enough context to be able to know which table it is

Exactly! `where` doesn't have that information available statically, which is required for the field to be available. This should work if you make `updateBatch` generic over the table:

```dart
  void updateBatch<T extends Table, D>(TableInfo<T, D> table, List<Insertable<D>> rows,
      Expression<bool> Function(T table) where) async {
    await _db.batch((batch) async {
      for (final row in rows) {
        batch.update(table, row, where: where);
      }
    });
  }
```

**Reply by piperun** [reply]
THanks for the fast reply! I'll try it out tomorrow!
